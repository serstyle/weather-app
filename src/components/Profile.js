import React, {Component} from 'react'
import { Redirect } from 'react-router-dom'

class Profile extends Component{
	state = {
		email: this.props.user.email,
		name: this.props.user.name,
		city: this.props.user.city,
		isAuth: true
	}


	componentDidMount(){
		fetch('https://weathers-server.herokuapp.com/profile', {
			method:'post',
			headers:{'Content-Type': 'application/json'},
			body: JSON.stringify({
				email:this.state.email
			})
		})
		.then(data => data.json())
		.then(user => {
			if(!user.email){
				return this.setState({isAuth:false})
			}
		})
		.catch(err => this.setState({isAuth:false}))

	}

	handleChange = (e) =>{
		this.setState({
			[e.target.id]: e.target.value
		})
	}

	handleSumbit = (e) => {
		e.preventDefault()
		const { email, city, name} = this.state
		fetch('https://weathers-server.herokuapp.com/profile', {
			method:'put',
			headers:{'Content-Type': 'application/json'},
			body: JSON.stringify({
				name,
				email,
				city
			})
		})
		.then(data => data.json())
		.then(user => {
			this.props.getUser(user)
			this.setState({isAuth:false})
		})
	}

	render(){
		if(!this.state.isAuth){
			return(<Redirect to='/weather-app' />)}


		return(
		<div className='container row'>
			<h1>Your Profile : </h1>
			<form onSubmit={this.handleSumbit} className='col s12'>
			      <div className="row">
			        <div className="input-field col s12">
			          <label className='active' htmlFor="name">Name</label>
			          <input onChange={this.handleChange} required id="name" value={this.state.name} type="text" className="validate" />
			        </div>
			      </div>
			      <div className="row">
			        <div className="input-field col s12">
			          <label className='active' htmlFor="city">City</label>
			          <input onChange={this.handleChange} value={this.state.city} required id="city" 
			      		type="text"/*don t forget to add city choice dont let user choose*/
			      		className="validate" />
			        </div>
			      </div>
			      <button className='waves-effect waves-light btn'> Submit </button>
		      </form>
	      </div>
			)
	}
}

export default Profile;