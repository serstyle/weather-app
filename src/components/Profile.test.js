import React from 'react'
import {shallow} from 'enzyme'
import Profile from './Profile'

const mockUser = {
    email:'', name:'name1', city:'city1'
}

const mockGetUser = jest.fn()


describe('test function', ()=>{
    describe('should handleChange', () => {
        let wrap = shallow(<Profile user={mockUser} getUser={mockGetUser}/>)
        it('onChange Input name', ()=>{
            const nameMock = wrap.state().name
            const event = {target: {id: "name", value: "name2"}};
            wrap.find('#name').simulate('change', event)
            expect(wrap.state().name).not.toBe(nameMock)
            expect(wrap.state().name).toBe('name2')
        })
        it('onChange Input city', ()=>{
            const cityMock = wrap.state().city
            const event = {target: {id: "city", value: "city2"}};
            wrap.find('#city').simulate('change', event)
            expect(wrap.state().city).not.toBe(cityMock)
            expect(wrap.state().city).toBe('city2')
        })
    })

    describe('should componentdidmount',()=>{
        it('should err', (done) => {
            const mockFailResponse = {}
            const mockJsonPromise = Promise.reject(mockFailResponse)
            const mockFetchPromise = Promise.reject({
                json: ()=> mockJsonPromise
            })
            jest.spyOn(global, 'fetch').mockImplementation(()=>mockFetchPromise)
            
            let wrap = shallow(<Profile user={mockUser} getUser={mockGetUser}/>)
            
            expect(global.fetch).toHaveBeenCalledTimes(1);
            expect(global.fetch).toHaveBeenCalledWith('https://weathers-server.herokuapp.com/profile', {
                method:'post',
                headers:{'Content-Type': 'application/json'},
                body: JSON.stringify({
                    email:wrap.state().email
                })
            })
            process.nextTick(() => { 
                expect(wrap.state()).toEqual({
                    email:'',
                    name:'name1',
                    city:'city1',
                    isAuth:false,
                    isSubmit:false
                });
                global.fetch.mockClear();
                done()
            });
        })
        
        it('should success', (done) => {
            const mockSuccessResponse = {}
            const mockJsonPromise = Promise.resolve(mockSuccessResponse)
            const mockFetchPromise = Promise.resolve({
                json: ()=> mockJsonPromise
            })
            jest.spyOn(global, 'fetch').mockImplementation(()=>mockFetchPromise)
    
            let wrap = shallow(<Profile user={mockUser} getUser={mockGetUser}/>)
            
            expect(global.fetch).toHaveBeenCalledTimes(1);
            expect(global.fetch).toHaveBeenCalledWith('https://weathers-server.herokuapp.com/profile', {
                method:'post',
                headers:{'Content-Type': 'application/json'},
                body: JSON.stringify({
                    email:wrap.state().email
                })
            })
            process.nextTick(() => { 
                expect(wrap.state()).toEqual({
                    email:'',
                    name:'name1',
                    city:'city1',
                    isAuth:false,
                    isSubmit:false
                });
                global.fetch.mockClear();
                done()
            });
        })  
    })  

    it('should handlesubmit', (done) => {
        let wrap = shallow(<Profile user={mockUser} getUser={mockGetUser}/>) 
        const mockSuccessResponse = {}
        const mockJsonPromise = Promise.resolve(mockSuccessResponse)
        const mockFetchPromise = Promise.resolve({
            json: ()=> mockJsonPromise
        })
        jest.spyOn(global, 'fetch').mockImplementation(()=>mockFetchPromise)

        wrap.setState({email:'test@test'})
        global.fetch.mockClear();
        wrap.find('form').simulate('submit', {
            preventDefault: () => {}
        })
        


        expect(global.fetch)
        .toHaveBeenCalledTimes(1);
        expect(global.fetch).toHaveBeenCalledWith('https://weathers-server.herokuapp.com/profile', {
            method:'put',
            headers:{'Content-Type': 'application/json'},
            body: JSON.stringify({
                name: wrap.state().name,
                email: wrap.state().email,
                city: wrap.state().city
            })
        })

        process.nextTick(() => { 
            wrap.setState({isAuth:true})
            expect(wrap.state()).toEqual({
                email:'test@test',
                name:'name1',
                city:'city1',
                isAuth:true,
                isSubmit:true
            });
            global.fetch.mockClear();
            done();
        });
    })
    

})

describe('snapshot with isAuth true or false', ()=>{
    let wrap = shallow(<Profile user={mockUser} getUser={mockGetUser}/>)
    it('should snapshot with isAuth true', () => {
      expect(wrap).toMatchSnapshot()
    })
    it('should snapshot with isAuth false', () => {
      wrap.setState({isAuth:false, isSubmit:true})
      expect(wrap).toMatchSnapshot()
    })
    
})