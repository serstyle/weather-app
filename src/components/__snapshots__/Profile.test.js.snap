// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot with isAuth true or false should snapshot with isAuth false 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Profile
    getUser={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {},
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    user={
      Object {
        "city": "city1",
        "email": "",
        "name": "name1",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "push": false,
      "to": "/",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "push": false,
        "to": "/",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`snapshot with isAuth true or false should snapshot with isAuth true 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Profile
    getUser={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {},
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    user={
      Object {
        "city": "city1",
        "email": "",
        "name": "name1",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          Your Profile : 
        </h1>,
        <form
          className="col s12"
          onSubmit={[Function]}
        >
          <div
            className="row"
          >
            <div
              className="input-field col s12"
            >
              <label
                className="active"
                htmlFor="name"
              >
                Name
              </label>
              <input
                className="validate"
                id="name"
                onChange={[Function]}
                required={true}
                type="text"
                value="name1"
              />
            </div>
          </div>
          <div
            className="row"
          >
            <div
              className="input-field col s12"
            >
              <label
                className="active"
                htmlFor="city"
              >
                City
              </label>
              <input
                className="validate"
                id="city"
                onChange={[Function]}
                required={true}
                type="text"
                value="city1"
              />
            </div>
          </div>
          <button
            className="waves-effect waves-light btn"
          >
             Submit 
          </button>
        </form>,
      ],
      "className": "container row",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Your Profile : ",
        },
        "ref": null,
        "rendered": "Your Profile : ",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="row"
            >
              <div
                className="input-field col s12"
              >
                <label
                  className="active"
                  htmlFor="name"
                >
                  Name
                </label>
                <input
                  className="validate"
                  id="name"
                  onChange={[Function]}
                  required={true}
                  type="text"
                  value="name1"
                />
              </div>
            </div>,
            <div
              className="row"
            >
              <div
                className="input-field col s12"
              >
                <label
                  className="active"
                  htmlFor="city"
                >
                  City
                </label>
                <input
                  className="validate"
                  id="city"
                  onChange={[Function]}
                  required={true}
                  type="text"
                  value="city1"
                />
              </div>
            </div>,
            <button
              className="waves-effect waves-light btn"
            >
               Submit 
            </button>,
          ],
          "className": "col s12",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <div
                className="input-field col s12"
              >
                <label
                  className="active"
                  htmlFor="name"
                >
                  Name
                </label>
                <input
                  className="validate"
                  id="name"
                  onChange={[Function]}
                  required={true}
                  type="text"
                  value="name1"
                />
              </div>,
              "className": "row",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <label
                    className="active"
                    htmlFor="name"
                  >
                    Name
                  </label>,
                  <input
                    className="validate"
                    id="name"
                    onChange={[Function]}
                    required={true}
                    type="text"
                    value="name1"
                  />,
                ],
                "className": "input-field col s12",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Name",
                    "className": "active",
                    "htmlFor": "name",
                  },
                  "ref": null,
                  "rendered": "Name",
                  "type": "label",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "validate",
                    "id": "name",
                    "onChange": [Function],
                    "required": true,
                    "type": "text",
                    "value": "name1",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
              ],
              "type": "div",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <div
                className="input-field col s12"
              >
                <label
                  className="active"
                  htmlFor="city"
                >
                  City
                </label>
                <input
                  className="validate"
                  id="city"
                  onChange={[Function]}
                  required={true}
                  type="text"
                  value="city1"
                />
              </div>,
              "className": "row",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <label
                    className="active"
                    htmlFor="city"
                  >
                    City
                  </label>,
                  <input
                    className="validate"
                    id="city"
                    onChange={[Function]}
                    required={true}
                    type="text"
                    value="city1"
                  />,
                ],
                "className": "input-field col s12",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "City",
                    "className": "active",
                    "htmlFor": "city",
                  },
                  "ref": null,
                  "rendered": "City",
                  "type": "label",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "validate",
                    "id": "city",
                    "onChange": [Function],
                    "required": true,
                    "type": "text",
                    "value": "city1",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
              ],
              "type": "div",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": " Submit ",
              "className": "waves-effect waves-light btn",
            },
            "ref": null,
            "rendered": " Submit ",
            "type": "button",
          },
        ],
        "type": "form",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Your Profile : 
          </h1>,
          <form
            className="col s12"
            onSubmit={[Function]}
          >
            <div
              className="row"
            >
              <div
                className="input-field col s12"
              >
                <label
                  className="active"
                  htmlFor="name"
                >
                  Name
                </label>
                <input
                  className="validate"
                  id="name"
                  onChange={[Function]}
                  required={true}
                  type="text"
                  value="name1"
                />
              </div>
            </div>
            <div
              className="row"
            >
              <div
                className="input-field col s12"
              >
                <label
                  className="active"
                  htmlFor="city"
                >
                  City
                </label>
                <input
                  className="validate"
                  id="city"
                  onChange={[Function]}
                  required={true}
                  type="text"
                  value="city1"
                />
              </div>
            </div>
            <button
              className="waves-effect waves-light btn"
            >
               Submit 
            </button>
          </form>,
        ],
        "className": "container row",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Your Profile : ",
          },
          "ref": null,
          "rendered": "Your Profile : ",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="row"
              >
                <div
                  className="input-field col s12"
                >
                  <label
                    className="active"
                    htmlFor="name"
                  >
                    Name
                  </label>
                  <input
                    className="validate"
                    id="name"
                    onChange={[Function]}
                    required={true}
                    type="text"
                    value="name1"
                  />
                </div>
              </div>,
              <div
                className="row"
              >
                <div
                  className="input-field col s12"
                >
                  <label
                    className="active"
                    htmlFor="city"
                  >
                    City
                  </label>
                  <input
                    className="validate"
                    id="city"
                    onChange={[Function]}
                    required={true}
                    type="text"
                    value="city1"
                  />
                </div>
              </div>,
              <button
                className="waves-effect waves-light btn"
              >
                 Submit 
              </button>,
            ],
            "className": "col s12",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <div
                  className="input-field col s12"
                >
                  <label
                    className="active"
                    htmlFor="name"
                  >
                    Name
                  </label>
                  <input
                    className="validate"
                    id="name"
                    onChange={[Function]}
                    required={true}
                    type="text"
                    value="name1"
                  />
                </div>,
                "className": "row",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <label
                      className="active"
                      htmlFor="name"
                    >
                      Name
                    </label>,
                    <input
                      className="validate"
                      id="name"
                      onChange={[Function]}
                      required={true}
                      type="text"
                      value="name1"
                    />,
                  ],
                  "className": "input-field col s12",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Name",
                      "className": "active",
                      "htmlFor": "name",
                    },
                    "ref": null,
                    "rendered": "Name",
                    "type": "label",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "className": "validate",
                      "id": "name",
                      "onChange": [Function],
                      "required": true,
                      "type": "text",
                      "value": "name1",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "input",
                  },
                ],
                "type": "div",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <div
                  className="input-field col s12"
                >
                  <label
                    className="active"
                    htmlFor="city"
                  >
                    City
                  </label>
                  <input
                    className="validate"
                    id="city"
                    onChange={[Function]}
                    required={true}
                    type="text"
                    value="city1"
                  />
                </div>,
                "className": "row",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <label
                      className="active"
                      htmlFor="city"
                    >
                      City
                    </label>,
                    <input
                      className="validate"
                      id="city"
                      onChange={[Function]}
                      required={true}
                      type="text"
                      value="city1"
                    />,
                  ],
                  "className": "input-field col s12",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "City",
                      "className": "active",
                      "htmlFor": "city",
                    },
                    "ref": null,
                    "rendered": "City",
                    "type": "label",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "className": "validate",
                      "id": "city",
                      "onChange": [Function],
                      "required": true,
                      "type": "text",
                      "value": "city1",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "input",
                  },
                ],
                "type": "div",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": " Submit ",
                "className": "waves-effect waves-light btn",
              },
              "ref": null,
              "rendered": " Submit ",
              "type": "button",
            },
          ],
          "type": "form",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
